### DISCUSSION TITLE
Should copy() still offer the option to replace the data?

### DISCUSSION THREAD

**Discussion by ESadek-MO** [SciTools/iris#6265] [General]
# See #6303.
During my work on #6253, I've stumbled into issues with DataManager.copy(). 

It seems that the current behavior, which allows you to replace the data at the same time as copying, though useful, doesn't really mirror how other packages use copy, and doesn't fit the new "contract" that 6253 has created around None. 

During discussions around this, the idea was floated of separating out the replacing behavior, for clarity sake and constistency.

This would no doubt be a breaking change, and therefore would have to go into the next major release.

**Comment by rcomer** [comment]
I use this a lot.  Is there a proposal for an alternative API which would achieve the same outcome?

**Reply by ESadek-MO** [reply]
There would be, yeah. 

This was written hurriedly before I left for the night, so there was something for today's peloton.

To add a bit more detail, the idea would be to separate out the `replace` functionality. So you'd have cube.copy(), which would only create a copy, and cube.replace() (or something!), which would copy and replace the data. 

The issue with the current way is that 
a) it seems to be unique to Iris, whereas in other packages, copy is just copy

b) with the introduction of dataless cubes, `None` is now essentially a **value meaning nothing**, as opposed to just meaning the absence of anything. Now, `iris.cube.Cube(data=None, shape=...)` is a dataless cube. So to have `my_cube.copy(data=None)` meaning don't replace data is confusing.

**Reply by bjlittle** [reply]
The subtle but important point above is that `data` in `iris.cube.Cube` will be a **kwarg** not an arg.

Also, `Cube.copy` will **always** return the same type of cube i.e., if you give it a cube with data then it will always return a duplicate cube with data, and if you give it a cube without data then it will always return a duplicate cube without data, which is a cleaner, simpler contract.

You can create a dataless cube as follows:

- `iris.cube.Cube(shape=(...))` or explicitly `iris.cube.Cube(data=None, shape=(...))`
- `cube.data = None`
- `cube.replace()` or `cube.replace(data=None)`

**Comment by trexfeathers** [comment]
I love this feature. Apologies!

**Comment by pp-mo** [comment]
From @scitools/peloton : basically agreed with this, for Iris v4 future.
Think the ```cube.replace()``` sounds a bit in-placey, another more explicitly functional name might be preferable (? "replaced" ?)

Lots to add here 
  * what about other components, e.g. dataless coords ?
  * what about **operations** on dataless cubes (we need a list of future possibles)
